src/vidpf.rs

```diff
@@ -493,43 +531,128 @@ pub struct VidpfPublicShare<W: VidpfValue> {
 }
 
 impl<W: VidpfValue> Encode for VidpfPublicShare<W> {
-    fn encode(&self, _bytes: &mut Vec<u8>) -> Result<(), CodecError> {
-        todo!()
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        // Control bits need to be written within each byte in LSB-to-MSB order, and assigned into
+        // bytes in big-endian order. Thus, the first four levels will have their control bits
+        // encoded in the last byte, and the last levels will have their control bits encoded in the
+        // first byte.
+        let mut control_bits: BitVec<u8, Lsb0> = BitVec::with_capacity(self.cw.len() * 2);
+        for correction_words in self.cw.iter() {
+            control_bits.extend(
+                [
+                    bool::from(correction_words.left_control_bit),
+                    bool::from(correction_words.right_control_bit),
+                ]
+                .iter(),
+            );
+        }
+        control_bits.set_uninitialized(false);
+        let mut packed_control = control_bits.into_vec();
+        bytes.append(&mut packed_control);
+
+        for correction_words in self.cw.iter() {
+            Seed(correction_words.seed).encode(bytes)?;
+            correction_words.weight.encode(bytes)?;
+        }
+
+        for proof in &self.cs {
+            bytes.extend_from_slice(proof);
+        }
+        Ok(())
     }
 
     fn encoded_len(&self) -> Option<usize> {
-        todo!()
+        let control_bits_count = (self.cw.len()) * 2;
+        let mut len = (control_bits_count + 7) / 8 + (self.cw.len()) * 16;
```

Comment:
I'm 99% sure I did not write this but it makes sense to me.

Reply:
Actually, no, it's just hardcoded throughout and needs to be 16 for use with both AES and TurboShake. I tried parameterizing it but this would involve rewriting some of the xof code and I think that's best left out of this PR